### PR TITLE
[ARCTIC-957] Arctic Flink adds type validation for the timestamp field defining watermark

### LINKED ISSUE #957: [Bug]: Flink sets watermark on Arctic table fields, but ArrayIndexOutOfBoundsException occurs when reading data
### What happened?

For an Arctic table, setting watermark on the time field of the table via Flink and doing a select operation on the table will result in an ArrayIndexOutOfBoundsException error.

### Affects Versions

master

### What engines are you seeing the problem on?

Flink

### How to reproduce

First create an Arctic table, where the field `ts` is a timestamp type field, and set the watermark based on the `ts` field; then do a GROUP BY operation based on this Arctic table and select the target field list, which will trigger the ArrayIndexOutOfBoundsException

### Relevant log output

```shell
Caused by: java.lang.ArrayIndexOutOfBoundsException: 4
	at com.netease.arctic.shaded.org.apache.iceberg.relocated.com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:75)
	at java.util.stream.IntPipeline$4$1.accept(IntPipeline.java:250)
	at java.util.Spliterators$IntArraySpliterator.forEachRemaining(Spliterators.java:1032)
	at java.util.Spliterator$OfInt.forEachRemaining(Spliterator.java:693)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
	at com.netease.arctic.flink.table.ArcticDynamicSource.applyProjection(ArcticDynamicSource.java:252)
	at org.apache.flink.table.planner.plan.rules.logical.PushProjectIntoTableSourceScanRule.onMatch(PushProjectIntoTableSourceScanRule.java:147)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:229)
```

### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by lklhdu**
fix #957 
## Why are the changes needed?

The Arctic Flink currently only supports timestamp(6) precision for timestamp types, however, Flink does not support defining watermarks on timestamp(6) fields and lacks syntax checks for timestamp types, which should be added.

## Brief change log

  - *Add type validation for the watermark definition*
  - *The Flink 1.12 version is affected.*

## How was this patch tested?
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / no) no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) not documented

### CODE REVIEW

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java [hunk 1]
```diff
@@ -184,4 +189,23 @@ public static void addPrimaryKey(
...
+      throw new ValidationException(
+        String.format(
+          "Invalid data type of expression for watermark definition. " +
+            "The field must be of type TIMESTAMP(p)," +
+            " the supported precision 'p' is from 0 to 3, but the watermark expression type is %s",
+          watermarkType));
+    }
+  }
+
+  public static boolean canBeTimeAttributeType(LogicalType logicalType) {
```
Comment (YesOrNo828) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java:203]: The flink-table-runtime module of the Flink project has provided [the util method](https://github.com/apache/flink/blob/master/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/TypeCheckUtils.java#L85) to distinguish whether the logical type variable is Timestamp or not.

**Review by YesOrNo828 — COMMENTED** [review]
@lklhdu  Thanks for your contribution. The change looks good. There is a util method already exists in the Flink project for the same functionality. I think you can remove this `canBeTimeAttributeType` method.

**Comment by YesOrNo828** [comment]
And I wanna know if Flink1.14/1.15 module can run correctly. Could you do this verification? @lklhdu

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java (hunk 1 above)
Comment (lklhdu) [reply flink/v1.12/flink/src/main/java/com/netease/arctic/flink/FlinkSchemaUtil.java]: Thanks for your review. I have made changes here.

**Comment by lklhdu** [comment]
> And I wanna know if Flink1.14/1.15 module can run correctly. Could you do this verification? @lklhdu

I run the TestWatermark unit test in the Flink1.14/1.15 module and find that due to the Flink version update, under Flink version 1.14/1.15, the Flink-table-api module adds validation for watermark fields and throws the same exception when running the following incorrect setting watermark way, and this part of the validation does not exist under Flink version 1.12:
```
create table user_tb (
    WATERMARK FOR ts AS ts - INTERVAL '5' SECOND
) like arctic_catalog.test.watermark_tb;
```
and run the unit test will throw this exception:
```
Caused by: org.apache.flink.table.api.ValidationException: Invalid data type of time field for watermark definition. The field must be of type TIMESTAMP(p) or TIMESTAMP_LTZ(p), the supported precision 'p' is from 0 to 3, but the time field type is TIMESTAMP(6)
	at org.apache.flink.table.catalog.DefaultSchemaResolver.validateTimeColumn(DefaultSchemaResolver.java:263)
	at org.apache.flink.table.catalog.DefaultSchemaResolver.resolveWatermarkSpecs(DefaultSchemaResolver.java:218)
	at org.apache.flink.table.catalog.DefaultSchemaResolver.resolve(DefaultSchemaResolver.java:84)
	at org.apache.flink.table.api.Schema.resolve(Schema.java:121)
        ...
```

**Comment by YesOrNo828** [comment]
LGTM, I'll merge into the master.

**Review by YesOrNo828 — APPROVED** [review]

**Pull request merged**